# Мини‑таск 1: Подготовка датасета для анализа активности молекул

### Импорт необходимых библиотек

In [206]:
import pandas as pd
import numpy as np
from rdkit import Chem

### Загрузка данных

При скачивании данных сразу отфильтровали записи, у которых тип активности IC50

In [207]:
df = pd.read_csv("data.csv", sep=';')

Посмотрим на первые строки

In [208]:
df.head()

Molecule ChEMBL ID Molecule Name  Molecule Max Phase  Molecular Weight  \
0        CHEMBL70104           NaN                 NaN            364.46   
1        CHEMBL96098           NaN                 NaN            372.88   
2        CHEMBL97179           NaN                 NaN            413.90   
3            CHEMBL6  INDOMETHACIN                 4.0            357.79   
4       CHEMBL355781           NaN                 NaN            390.46   

   #RO5 Violations  AlogP  Compound Key  \
0              0.0   3.49             3   
1              0.0   4.43            34   
2              1.0   5.19             5   
3              0.0   3.93  Indomethacin   
4              0.0   4.25            21   

                                              Smiles Standard Type  \
0    CC1(C)OC(=O)C(OC2CCCCC2)=C1c1ccc(S(C)(=O)=O)cc1          IC50   
1     CCc1ccc(-c2ncc(Cl)cc2-c2ccc(S(C)(=O)=O)cc2)cn1          IC50   
2  CCCCOC(=O)Cc1c(C)n(C(=O)c2ccc(Cl)cc2)c2ccc(OC)...          IC50   
3     COc1ccc2c(c1)c(CC(=O)O)c(C)n2C(=O)c1ccc(Cl)cc1          IC50   
4  CS(=O)(=O)c1ccc(-c2csc(CC(=O)O)c2-c2ccc(F)cc2)cc1          IC50   

  Standard Relation  ...  Document ChEMBL ID Source ID     Source Description  \
0               '='  ...       CHEMBL1131821         1  Scientific Literature   
1               '='  ...       CHEMBL1130609         1  Scientific Literature   
2               '='  ...       CHEMBL1133457         1  Scientific Literature   
3               '<'  ...       CHEMBL1131754         1  Scientific Literature   
4               '>'  ...       CHEMBL1128908         1  Scientific Literature   

       Document Journal Document Year Cell ChEMBL ID  Properties  Action Type  \
0  Bioorg Med Chem Lett        1999.0            NaN         NaN          NaN   
1  Bioorg Med Chem Lett        1998.0            NaN         NaN          NaN   
2            J Med Chem        2000.0            NaN         NaN          NaN   
3  Bioorg Med Chem Lett        1999.0            NaN         NaN          NaN   
4  Bioorg Med Chem Lett        1996.0  CHEMBL3308860         NaN          NaN   

   Standard Text Value   Value  
0                  NaN    0.04  
1                  NaN    1.70  
2                  NaN    0.05  
3                  NaN  200.00  
4                  NaN   10.00  

[5 rows x 48 columns]

Размер датасета

In [209]:
df.shape

(7979, 48)

Тип колонок

In [210]:
df.dtypes

Molecule ChEMBL ID             object
Molecule Name                  object
Molecule Max Phase            float64
Molecular Weight              float64
#RO5 Violations               float64
AlogP                         float64
Compound Key                   object
Smiles                         object
Standard Type                  object
Standard Relation              object
Standard Value                float64
Standard Units                 object
pChEMBL Value                 float64
Data Validity Comment          object
Comment                        object
Uo Units                       object
Ligand Efficiency BEI         float64
Ligand Efficiency LE          float64
Ligand Efficiency LLE         float64
Ligand Efficiency SEI         float64
Potential Duplicate             int64
Assay ChEMBL ID                object
Assay Description              object
Assay Type                     object
BAO Format ID                  object
BAO Label                      object
Assay Organi

Информацию о датасете

In [211]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7979 entries, 0 to 7978
Data columns (total 48 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Molecule ChEMBL ID          7979 non-null   object 
 1   Molecule Name               1782 non-null   object 
 2   Molecule Max Phase          1462 non-null   float64
 3   Molecular Weight            7966 non-null   float64
 4   #RO5 Violations             7873 non-null   float64
 5   AlogP                       7873 non-null   float64
 6   Compound Key                7979 non-null   object 
 7   Smiles                      7945 non-null   object 
 8   Standard Type               7979 non-null   object 
 9   Standard Relation           6978 non-null   object 
 10  Standard Value              6978 non-null   float64
 11  Standard Units              6997 non-null   object 
 12  pChEMBL Value               5653 non-null   float64
 13  Data Validity Comment       340 n

Статистическое описание данных

In [212]:
df.describe()

Molecule Max Phase  Molecular Weight  #RO5 Violations        AlogP  \
count         1462.000000       7966.000000      7873.000000  7873.000000   
mean             3.705540        386.010060         0.268767     3.755918   
std              0.880034        129.600237         0.563463     1.611136   
min             -1.000000         32.050000         0.000000   -13.050000   
25%              4.000000        329.372500         0.000000     2.940000   
50%              4.000000        379.435000         0.000000     3.780000   
75%              4.000000        423.440000         0.000000     4.590000   
max              4.000000       2324.380000         4.000000    13.310000   

       Standard Value  pChEMBL Value  Ligand Efficiency BEI  \
count    6.978000e+03    5653.000000            5563.000000   
mean     1.043837e+06       6.332271              17.005918   
std      1.957618e+07       1.141168               4.400239   
min      0.000000e+00       4.000000               2.230000   
25%      1.000000e+02       5.500000              14.070000   
50%      9.800000e+02       6.300000              16.750000   
75%      1.000000e+04       7.150000              19.625000   
max      8.333357e+08      10.700000              53.820000   

       Ligand Efficiency LE  Ligand Efficiency LLE  Ligand Efficiency SEI  \
count           5533.000000            5533.000000            5531.000000   
mean               0.340481               2.486374              10.119599   
std                0.087585               1.650316               6.565549   
min                0.100000              -6.880000               1.560000   
25%                0.280000               1.520000               6.810000   
50%                0.340000               2.600000               8.990000   
75%                0.400000               3.610000              11.525000   
max                0.990000              10.720000             166.600000   

       Potential Duplicate  Assay Variant Accession  Assay Variant Mutation  \
count          7979.000000                      0.0                     0.0   
mean              0.069934                      NaN                     NaN   
std               0.255051                      NaN                     NaN   
min               0.000000                      NaN                     NaN   
25%               0.000000                      NaN                     NaN   
50%               0.000000                      NaN                     NaN   
75%               0.000000                      NaN                     NaN   
max               1.000000                      NaN                     NaN   

         Source ID  Document Year  Standard Text Value          Value  
count  7979.000000    7108.000000                  0.0    6978.000000  
mean      2.989222    2009.073016                  NaN     605.346449  
std       5.948862       8.228747                  NaN    7105.585059  
min       1.000000    1994.000000                  NaN      -8.220000  
25%       1.000000    2001.000000                  NaN       0.281000  
50%       1.000000    2010.000000                  NaN       2.800000  
75%       1.000000    2016.000000                  NaN      15.425000  
max      65.000000    2024.000000                  NaN  100000.000000

Пропущенные значения

In [213]:
df.isnull().sum()

Molecule ChEMBL ID               0
Molecule Name                 6197
Molecule Max Phase            6517
Molecular Weight                13
#RO5 Violations                106
AlogP                          106
Compound Key                     0
Smiles                          34
Standard Type                    0
Standard Relation             1001
Standard Value                1001
Standard Units                 982
pChEMBL Value                 2326
Data Validity Comment         7639
Comment                       6890
Uo Units                       982
Ligand Efficiency BEI         2416
Ligand Efficiency LE          2446
Ligand Efficiency LLE         2446
Ligand Efficiency SEI         2448
Potential Duplicate              0
Assay ChEMBL ID                  0
Assay Description                0
Assay Type                       0
BAO Format ID                    0
BAO Label                        0
Assay Organism                3665
Assay Tissue ChEMBL ID        7018
Assay Tissue Name   

Хоть и скачали датасет с фильтрацией, но проверим у всех ли тип активности IC50

In [214]:
df = df[df['Standard Type'] == 'IC50'].copy()

In [215]:
df.shape

(7979, 48)

### Отбор релевантных колонок

|  № | Колонка                    | Тип     | Заполнено | Описание                                                 |
| -: |:---------------------------|:--------| --------: |:---------------------------------------------------------|
|  0 | Molecule ChEMBL ID         | object  |      7979 | Уникальный id молекулы в ChEMBL                          |
|  1 | Molecule Name              | object  |      1782 | Название соединения                                      |
|  2 | Molecule Max Phase         | float64 |      1462 | Максимальная фаза клинических испытаний                  |
|  3 | Molecular Weight           | float64 |      7966 | Молекулярная масса                                       |
|  4 | #RO5 Violations            | float64 |      7873 | Количество нарушений правила пяти Липински               |
|  5 | AlogP                      | float64 |      7873 | Гидрофобность молекул                                    |
|  6 | Compound Key               | object  |      7979 | Альтернативный id соединения                             |
|  7 | Smiles                     | object  |      7945 | Представление структуры в SMILES                         |
|  8 | Standard Type              | object  |      7979 | Тип активности                                           |
|  9 | Standard Relation          | object  |      6978 | Знак отношения (=, <, >, >=)                             |
| 10 | Standard Value             | float64 |      6978 | Значение активности                                      |
| 11 | Standard Units             | object  |      6997 | Единицы измерения активности                             |
| 12 | pChEMBL Value              | float64 |      5653 | –log10(Standard Value)                                   |
| 13 | Data Validity Comment      | object  |       340 | Комментарий к достоверности данных                       |
| 14 | Comment                    | object  |      1089 | Общий комментарий к эксперименту                         |
| 15 | Uo Units                   | object  |      6997 | Онтология единиц измерения                               |
| 16 | Ligand Efficiency BEI      | float64 |      5563 | Активность на массу молекулы                             |
| 17 | Ligand Efficiency LE       | float64 |      5533 | Эффективность связывания на тяжёлый атом                 |
| 18 | Ligand Efficiency LLE      | float64 |      5533 | Баланс активности и липофильности                        |
| 19 | Ligand Efficiency SEI      | float64 |      5531 | Активность на полярную поверхность молекулы              |
| 20 | Potential Duplicate        | int64   |      7979 | Флаг на наличие дубликата (0 или 1)                      |
| 21 | Assay ChEMBL ID            | object  |      7979 | Id биологического эксперимента                           |
| 22 | Assay Description          | object  |      7979 | Краткое описание эксперимента                            |
| 23 | Assay Type                 | object  |      7979 | Тип эксперимента (A - ADME, B — binding, F — functional) |
| 24 | BAO Format ID              | object  |      7979 | BAO id формата эксперимента                              |
| 25 | BAO Label                  | object  |      7979 | Название формата по BAO                                  |
| 26 | Assay Organism             | object  |      4314 | Организм, в котором проводился эксперимент               |
| 27 | Assay Tissue ChEMBL ID     | object  |       961 | Id ткани                                                 |
| 28 | Assay Tissue Name          | object  |       962 | Название ткани                                           |
| 29 | Assay Cell Type            | object  |      1669 | Тип клеток                                               |
| 30 | Assay Subcellular Fraction | object  |        24 | Подклеточная фракция                                     |
| 31 | Assay Parameters           | object  |         8 | Дополнительные параметры                                 |
| 32 | Assay Variant Accession    | float64 |         0 |                                                          |
| 33 | Assay Variant Mutation     | float64 |         0 |                                                          |
| 34 | Target ChEMBL ID           | object  |      7979 | Id мишени                                                |
| 35 | Target Name                | object  |      7979 | Название мишени                                          |
| 36 | Target Organism            | object  |      7979 | Организм мишени                                          |
| 37 | Target Type                | object  |      7979 | Тип мишени                                               |
| 38 | Document ChEMBL ID         | object  |      7979 | Id научного источника                                    |
| 39 | Source ID                  | int64   |      7979 | Внутренний id источника                                  |
| 40 | Source Description         | object  |      7979 | Тип источника                                            |
| 41 | Document Journal           | object  |      7011 | Название журнала                                         |
| 42 | Document Year              | float64 |      7108 | Год публикации                                           |
| 43 | Cell ChEMBL ID             | object  |      1388 | Id клеточной линии                                       |
| 44 | Properties                 | object  |       536 | Дополнительные свойства                                  |
| 45 | Action Type                | object  |       801 | Тип действия                                             |
| 46 | Standard Text Value        | float64 |         0 |                                                          |
| 47 | Value                      | float64 |      6978 | Альтернативное числовое значение активности              |


Выбираем релевантные колонки, которые отражают биологическую активность молекул или понадобятся для дальнейшей работы и также у которых нет много пропущенных значений

In [216]:
relevant_columns = [
    'Smiles',
    'Molecular Weight',
    '#RO5 Violations',
    'AlogP',
    'Standard Value',
    'Standard Units'
]

Сохраняем в новый датасет

In [217]:
df = df[relevant_columns].reset_index(drop=True)

Выведем первые строки нового датасета

In [218]:
df.head()

Smiles  Molecular Weight  \
0    CC1(C)OC(=O)C(OC2CCCCC2)=C1c1ccc(S(C)(=O)=O)cc1            364.46   
1     CCc1ccc(-c2ncc(Cl)cc2-c2ccc(S(C)(=O)=O)cc2)cn1            372.88   
2  CCCCOC(=O)Cc1c(C)n(C(=O)c2ccc(Cl)cc2)c2ccc(OC)...            413.90   
3     COc1ccc2c(c1)c(CC(=O)O)c(C)n2C(=O)c1ccc(Cl)cc1            357.79   
4  CS(=O)(=O)c1ccc(-c2csc(CC(=O)O)c2-c2ccc(F)cc2)cc1            390.46   

   #RO5 Violations  AlogP  Standard Value Standard Units  
0              0.0   3.49            40.0             nM  
1              0.0   4.43          1700.0             nM  
2              1.0   5.19            50.0             nM  
3              0.0   3.93           200.0             nM  
4              0.0   4.25         10000.0             nM

### Приведение значений активности к единому формату

Создаем функцию, которая приводит значения активности к нм

In [219]:
def convert_to_nM(row):
    val  = row['Standard Value']
    unit = row['Standard Units']
    mw   = row['Molecular Weight']

    try:
        v = float(val)
    except:
        return np.nan

    if pd.isna(unit):
        return np.nan

    u = unit.strip().lower()

    if u == 'nm':
        return v

    if u == 'um':
        return v * 1e3

    if u =='ug.ml-1':
        if pd.notna(mw) and mw > 0:
            return (v * 1e3) / mw
        else:
            return np.nan

    # все остальные (%, ug без объёма)
    return np.nan

Применяем функцию конвертирования и сохраняем в новую колонку

In [220]:
df['Activity'] = df.apply(convert_to_nM, axis=1)

In [221]:
df.head()

Smiles  Molecular Weight  \
0    CC1(C)OC(=O)C(OC2CCCCC2)=C1c1ccc(S(C)(=O)=O)cc1            364.46   
1     CCc1ccc(-c2ncc(Cl)cc2-c2ccc(S(C)(=O)=O)cc2)cn1            372.88   
2  CCCCOC(=O)Cc1c(C)n(C(=O)c2ccc(Cl)cc2)c2ccc(OC)...            413.90   
3     COc1ccc2c(c1)c(CC(=O)O)c(C)n2C(=O)c1ccc(Cl)cc1            357.79   
4  CS(=O)(=O)c1ccc(-c2csc(CC(=O)O)c2-c2ccc(F)cc2)cc1            390.46   

   #RO5 Violations  AlogP  Standard Value Standard Units  Activity  
0              0.0   3.49            40.0             nM      40.0  
1              0.0   4.43          1700.0             nM    1700.0  
2              1.0   5.19            50.0             nM      50.0  
3              0.0   3.93           200.0             nM     200.0  
4              0.0   4.25         10000.0             nM   10000.0

Посмотрим сколько строк было конвертировано

In [222]:
df['Activity'].notna().sum()

np.int64(6976)

### Обработка пропусков и дубликатов

Удаляем дубликаты

In [223]:
df = df.drop_duplicates().reset_index(drop=True)

Видим, что дубликатов нет

In [224]:
df.shape

(7371, 7)

Удаляем записи у которых нет расчитанной активности

In [225]:
df = df[df['Activity'].notna()].copy()

Посмотрим на размер

In [226]:
df.shape

(6393, 7)

### Проверка корректности SMILES

In [227]:
def is_valid_smiles(smiles):
    try:
        mol = Chem.MolFromSmiles(smiles)
        return mol is not None
    except:
        return False

Фильтрация по валидным SMILES

In [228]:
mask = df['Smiles'].apply(is_valid_smiles)
df = df[mask].reset_index(drop=True)

Смотрим на результат

In [229]:
df.shape

(6392, 7)

### Сохранение нового датасета

Удаляем колонку Standard Units, он больше не нужен

In [230]:
df.drop('Standard Units', axis=1, inplace=True)

In [231]:
df.to_csv("new_data.csv", index=False)